### DISCUSSION TITLE
ets and dets

### DISCUSSION THREAD

**Discussion by eko234** [gleam-lang/gleam#1717] [Questions & support]
Hi, I'm new to gleam and wanted to know how could I use ets and dets from it, I want to implement a little service and store low data volumes with it.

**Comment by inoas** [comment]
I'd suggest an opaque type either for key/value or for a whole record.

Here it is not persisted https://github.com/inoas/gleam-elixir-phoenix-liveview-counter/blob/main/src/counter.gleam but every increment/decrement/set_value/get_value could instead read from ets via FFI: https://gleam.run/book/tour/external-functions.html

Here is some example from stdlib for FFI:
- https://github.com/gleam-lang/stdlib/blob/main/src/gleam_stdlib.erl#L320-L322 define ffi
- https://github.com/gleam-lang/stdlib/blob/main/src/gleam_stdlib.erl#L11 export ffi
- https://github.com/gleam-lang/stdlib/blob/main/src/gleam/base.gleam#L34-L37 use ffi

**Comment by lpil** [comment]
Hello! You would need to design and implement API bindings for ETS and DETS using [external functions](https://gleam.run/book/tour/external-functions.html), none currently exist for Gleam.

If I wanted to store small volumes of data like that I would consider Khepri over DETS, it is more modern and in my opinion has a nicer interface.

https://github.com/rabbitmq/khepri

**Comment by eko234** [comment]
Thanks, right now I'm creating my own ffi bindings, it's been relatively easy, really liking the language so far.

**Reply by lpil** [reply]
Great to hear!

**Comment by mjsir911** [comment]
so, I've found & updated https://github.com/zunda-arrow/gts but apparently "the gts library has lots of problems" per #4987 . is there a decent alternative? 

https://github.com/gleam-lang/erlang/issues/39 asked about this as well

https://github.com/Michael-Mark-Edu/bravo seems like something? although it doesn't seem to support name-based lookups, which is what I need.

there's some mention of acknowledged desire for mutable state on https://gleam.run/frequently-asked-questions/#does-gleam-have-mutable-state and it mentions to ets, but it doesn't propose any way to actually *use* it other than building your own ffi bindings

**Reply by lpil** [reply]
If you search the package index for "ets" you'll find multiple options, and you can use the Erlang APIs directly too. What is best depends on your specific use case.

**Reply by mjsir911** [reply]
https://packages.gleam.run/?search=ETS
